File: src/main/org/nlogo/agent/TrigTables.java
package org.nlogo.agent;

public final strictfp class TrigTables {

  // this class is not instantiable
  private TrigTables() {
    throw new IllegalStateException();
  }

  /// cache of sines and cosines of integers 0-359

  // This is to speed up various operations on angles
  // that are integers.  Trigonometry is expensive!!
  // - ST 12/21/06, 12/24/06

  public static final double[] sin = new double[360];
  public static final double[] cos = new double[360];

  static {
    for (int i = 0; i < 360; i++) {
      double headingRadians = StrictMath.toRadians(i);
      double c = StrictMath.cos(headingRadians);
      double s = StrictMath.sin(headingRadians);
      if (StrictMath.abs(c) < org.nlogo.api.World.INFINITESIMAL) {
        c = 0;
      }
      if (StrictMath.abs(s) < org.nlogo.api.World.INFINITESIMAL) {
        s = 0;
      }
      sin[i] = s;
      cos[i] = c;
    }
  }

}

File: src/main/org/nlogo/agent/TrigTables.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

import org.nlogo.api.Numbers.Infinitesimal

@annotation.strictfp
/**
 * Cache of sines and cosines of integers 0-359.
 * Speed up various operations on angles that are integers.  Trigonometry is expensive!
 */
object TrigTables {
  val sin = Array.tabulate(360)(mySin)
  val cos = Array.tabulate(360)(myCos)
  private def mySin(n: Int): Double = {
    val radians = StrictMath.toRadians(n)
    val x = StrictMath.sin(radians)
    if (StrictMath.abs(x) < Infinitesimal)
      0
    else
      x
  }
  private def myCos(n: Int): Double = {
    val radians = StrictMath.toRadians(n)
    val x = StrictMath.cos(radians)
    if (StrictMath.abs(x) < Infinitesimal)
      0
    else
      x
  }
}
